[Issue 1976] Limit hot states kept in memory

mbaxter opened the pull request:
<!-- Thanks for sending a pull request! Please check out our contribution guidelines: -->
<!-- https://github.com/PegaSysEng/teku/blob/master/CONTRIBUTING.md -->

## PR Description
Limit the number of states that are kept in memory.  If a state that has been dropped from memory is later requested, regenerate the missing state.

## Fixed Issue(s)
Fixes #1976

mbaxter committed:
Limit the number of hot states kept in memory

mbaxter committed:
Revert "Simplify finalized checkpoint handling"

This reverts commit 478590d098a7ded563ef64d41cb1d51659a3fe04.

mbaxter committed:
Don't hold the read lock while regenerating states

Keep track of latest finalized block and state so we always have a
base state that allows us to rebuild any given a list of subsequent
blocks.

mbaxter committed:
Optimize state lookups when collecting finalized chain data

mbaxter committed:
Fix test - set finalized checkpoint that supersedes genesis

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/store/Store.java:
```diff
@@ -212,17 +233,12 @@ public SignedBeaconBlock getSignedBlock(Bytes32 blockRoot) {
+    if (block == null || state.isEmpty()) {
+      return Optional.empty();
     }
+    return Optional.of(new SignedBlockAndState(block, state.get()));
```
Would it be worth bailing out before attempting to recreate the state if `block == null`?  Performance-wise it won't make any significant difference but it feels weird to ask for something we already know we can't generate.  I think you could then just use a `map`:
```suggestion
    final SignedBeaconBlock block = getSignedBlock(blockRoot);
    if (block == null) {
      return Optional.empty();
    }
    return getOrGenerateBlockState(blockRoot).map(state -> new SignedBlockAndState(block, state));
```

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/store/Store.java:
```diff
@@ -247,51 +263,110 @@ public boolean containsBlock(Bytes32 blockRoot) {
+    final StateGenerator stateGenerator = new StateGenerator();
+    final Map<Bytes32, BeaconState> regeneratedStates =
+        stateGenerator.produceStatesForBlocks(
+            baseBlock.getRoot(), baseBlock.getState(), blocks.values());
```
This could be a memory issue for us - say we'd wound up with the last finalised epoch being 10 and had gone on importing blocks so that now we've only got states from epoch 20 onwards in memory.  Then we get a request for a state in epoch 19, we'd regenerate all the states from epoch 10 to 19 and they'd all be in memory at once in that map which would cause use to run out of memory.

I suspect we could just build up the list of blocks to import in slot order as we walk back to find the most recent state, then we can just iterate through them to regenerate the state.  Something like:
```
private Optional<BeaconState> getOrGenerateBlockState(final Bytes32 blockRoot) {
    Optional<BeaconState> state = getState(blockRoot);
    if (state.isPresent()) {
      return state;
    }
    final SignedBeaconBlock blockForState = getSignedBlock(blockRoot);
    if (blockForState == null) {
      // If we don't have the corresponding block, we can't possibly regenerate the state
      return Optional.empty();
    }

    // Accumulate blocks until we find our base state to build from
    SignedBlockAndState baseBlock = null;
    final List<SignedBeaconBlock> blocks = new ArrayList<>(); // in reverse slot order
    SignedBeaconBlock block = blockForState;
    while (block != null) {
      final Optional<BeaconState> blockState = getState(block.getRoot());
      if (blockState.isPresent()) {
        // We found a base state
        baseBlock = new SignedBlockAndState(block, blockState.get());
        break;
      }
      blocks.add(block);
      block = getSignedBlock(block.getParent_root());
    }

    if (baseBlock == null) {
      // If we haven't found a base state yet, we must have walked back to the latest finalized
      // block, check here for the base state
      final SignedBlockAndState finalizedBlock = getLatestFinalizedBlockAndState();
      final SignedBeaconBlock earliestBlock = blocks.get(0);
      if (!earliestBlock.getRoot().equals(finalizedBlock.getRoot())) {
        // We must have finalized a new block while processing and moved past our target root
        return Optional.empty();
      }
      baseBlock = finalizedBlock;
    }

    // Regenerate state
    BeaconState regeneratedState = baseBlock.getState();
    final StateGenerator stateGenerator = new StateGenerator();
    for (int i = blocks.size() - 1; i >= 0; i--) {
      regeneratedState = stateGenerator.processBlock(regeneratedState, blocks.get(i));
    }
    // Save regenerated state
    putState(blockRoot, regeneratedState);

    return Optional.of(regeneratedState);
  }
```

Though probably should move all the state reiteration to a new method in `StateGenerator` rather than doing it here. Need to double check the logic on checking if we've walked past the finalised state as well - I'd be slightly tempted to gather up the blocks to apply inside a read lock so we know the finalised state won't change, then once we have them release the lock and do the state transitions to generate the state.

ajsutton replied:
For a follow up ticket - we should ensure we can restart in the middle of a long period of non-finalisation too.  I think currently we'd regenerate all states and need to hold them in memory which would cause issues.

mbaxter replied:
> I'd be slightly tempted to gather up the blocks to apply inside a read lock so we know the finalised state won't change, then once we have them release the lock and do the state transitions to generate the state

We could do that, but when we come back around to block pruning, that could cause issues.  So, I think I'll leave it for now ...

ajsutton replied:
Yeah fair point.

mbaxter replied:
Reworked the `StateGenerator` so that we only ever process one state at a time and limit the number of states we keep in memory.

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/store/StoreTransactionUpdates.java:
```diff
@@ -123,24 +131,35 @@ public static StoreTransactionUpdates calculate(final Store baseStore, final Tra
+    // State collection is done separately to avoid repeatedly regenerating state by processing
+    // states in reverse-order.  See: Store.getOrGenerateBlockState
+    Collections.reverse(blocks);
+    return blocks.stream()
+        .map(
+            block -> {
+              final BeaconState state = tx.getBlockState(block.getRoot());
+              return new SignedBlockAndState(block, state);
+            })
+        .collect(Collectors.toMap(SignedBlockAndState::getRoot, Function.identity()));
```
Probably worth logging a ticket to revisit this in the future - it probably isn't smart to regenerate a bunch of states just to send them to disk (especially since if we're pruning they'll then be thrown away).  When our database is a bit more advanced we'll likely only store periodic state snapshots anyway so just not storing states we didn't have in memory seems like a good idea.

mbaxter replied:
Added a follow-up here: https://github.com/PegaSysEng/teku/issues/2101

mbaxter committed:
Make state generation logic more efficient

Don't build all states at once, instead process regenerated states one
at a time.

mbaxter committed:
Merge branch 'master' into issue-1976/limit-hot-states-kept-in-memory

mbaxter committed:
Exit getBlockAndState early if block is missing

mbaxter committed:
Clarify variable names

mbaxter commented on storage/src/main/java/tech/pegasys/teku/storage/store/StoreTransactionUpdates.java:
```diff
@@ -210,17 +229,24 @@ public void applyToStore(final Store store) {
     tx.genesis_time.ifPresent(value -> store.genesis_time = value);
     tx.justified_checkpoint.ifPresent(value -> store.justified_checkpoint = value);
-    tx.finalized_checkpoint.ifPresent(value -> store.finalized_checkpoint = value);
+    newFinalizedCheckpoint.ifPresent(value -> store.finalized_checkpoint = value.getCheckpoint());
```
For consistency, we're now only setting a new finalized checkpoint if the `tx` checkpoint made it through the epoch check (if the new finalized epoch is greater than the current finalized epoch): https://github.com/PegaSysEng/teku/blob/4963b9098e9f92d28dd72e58ae01f6023439b8e2/storage/src/main/java/tech/pegasys/teku/storage/store/StoreTransactionUpdates.java#L82-L84

mbaxter committed:
Clean up

mbaxter committed:
Simplify code

mbaxter committed:
Revert "Simplify code" - this form ends up caching too many states

This reverts commit acdbabed62fd6941c38abf3aa2d2b8c9937bbec4.

mbaxter commented on ethereum/core/src/main/java/tech/pegasys/teku/core/StateGenerator.java:
```diff
@@ -13,64 +13,136 @@
 
+/** Utility for regenerating block states given a block tree and a root state. */
 public class StateGenerator {
+  private static final int DEFAULT_STATE_CACHE_SIZE = 50;
```
Any opinions on a good default size here?  50 seems like a lot - maybe 25?

ajsutton replied:
I'd rather err on the side of too much cache since the performance cost of misses is pretty severe.  We keep 5 epochs worth of states in the actual store so I'd stick with 50.  We can tune down if we have memory pressure but given this is run at startup we have more headroom than usual anyway.

mbaxter commented on storage/src/main/java/tech/pegasys/teku/storage/store/Store.java:
```diff
@@ -247,51 +267,109 @@ public boolean containsBlock(Bytes32 blockRoot) {
+      baseBlock = finalizedBlock;
+    }
+
+    // Regenerate state
+    final BlockTree tree =
+        BlockTree.builder().rootBlock(baseBlock.getBlock()).blocks(blocks.values()).build();
+    final StateGenerator stateGenerator = new StateGenerator(tree, baseBlock.getState());
+    final BeaconState regeneratedState = stateGenerator.regenerateStateForBlock(blockRoot);
```
There's some unnecessary processing happening behind the scenes when we create the `BlockTree`.  I may try to simplify this some more, but I think it works okay as-is.

mbaxter committed:
Simplify list handling

mbaxter committed:
Merge branch 'master' into issue-1976/limit-hot-states-kept-in-memory

ajsutton commented on ethereum/core/src/main/java/tech/pegasys/teku/core/StateGenerator.java:
```diff
@@ -13,64 +13,137 @@
+    BeaconState prevState = baseState;
+    SignedBeaconBlock currentBlock = null;
+    BeaconState currentState = null;
+    for (int i = 0; i < blocks.size(); i++) {
```
nit: Given we're using indexed iteration anyway it's probably better to just iterate backwards through the list rather than using `Collections.reverse` on it.

ajsutton commented on ethereum/core/src/main/java/tech/pegasys/teku/core/StateGenerator.java:
```diff
@@ -13,64 +13,137 @@
+    for (int i = 0; i < blocks.size(); i++) {
+      currentBlock = blocks.get(i);
+      currentState = processBlock(prevState, currentBlock);
+      prevState = currentState;
```
It's a bit confusing that after this loop we wind up using `prevState` for the resulting state after applying the last block.  Would probably be simpler to just remove `prevState` and use only currentState so this winds up as:
```suggestion
      currentState = processBlock(currentState, currentBlock);
```

ajsutton replied:
Possibly just `state` rather than `currentState` even.

ajsutton commented on ethereum/core/src/main/java/tech/pegasys/teku/core/StateGenerator.java:
```diff
@@ -13,64 +13,137 @@
+          children.subList(1, children.size()).forEach(branchesToProcess::push);
+        }
+        // Continue processing the first child
+        branchBlock = children.size() > 0 ? children.get(0) : null;
```
nit: 
```suggestion
        branchBlock = !children.isEmpty() ? children.get(0) : null;
```

ajsutton commented on storage/src/testFixtures/java/tech/pegasys/teku/storage/store/StoreAssertions.java:
```diff
@@ -0,0 +1,25 @@
+
+public class StoreAssertions {
+
+  public static void assertStoresMatch(
```
nice.

ajsutton reviewed (approved):
LGTM.

mbaxter committed:
Make vars private, clean up code

mbaxter committed:
Merge branch 'master' into issue-1976/limit-hot-states-kept-in-memory
